src/persistence.rs

```diff
@@ -126,21 +126,46 @@ pub struct AtomicKeystoreStorage<'a, L: Ledger, Meta: Serialize + DeserializeOwn
     static_dirty: bool,
     dynamic_state: RollingLog<EncryptingResourceAdapter<KeystoreSnapshot<L>>>,
     dynamic_state_dirty: bool,
-    keystore_key_tree: KeyTree,
+    root_key_tree: KeyTree,
 }
 
 impl<'a, L: Ledger, Meta: Send + Serialize + DeserializeOwned + Clone + PartialEq>
     AtomicKeystoreStorage<'a, L, Meta>
 {
-    pub fn new<T>(
-        adapter: EncryptingResourceAdapter<T>,
+    pub fn new(
+        loader: &mut impl KeystoreLoader<L, Meta = Meta>,
         atomic_loader: &mut AtomicStoreLoader,
         file_fill_size: u64,
-        meta: Meta,
-        persisted_meta: RollingLog<BincodeLoadStore<Meta>>,
-        meta_dirty: bool,
-        key: KeyTree,
     ) -> Result<Self, KeystoreError<L>> {
+        // Load the metadata first so the loader can use it to generate the encryption key needed to
+        // read the rest of the data.
+        let mut persisted_meta = RollingLog::load(
+            atomic_loader,
+            BincodeLoadStore::<Meta>::default(),
+            "keystore_meta",
+            1024,
+        )?;
+        let (meta, key, meta_dirty) = match persisted_meta.load_latest() {
+            Ok(mut meta) => {
+                let old_meta = meta.clone();
+                let key = loader.load(&mut meta)?;
+
+                // Store the new metadata if the loader changed it
+                if meta != old_meta {
+                    persisted_meta.store_resource(&meta)?;
+                    (meta, key, true)
+                } else {
+                    (meta, key, false)
+                }
+            }
+            Err(_) => {
+                // If there is no persisted metadata, ask the loader to generate a new keystore.
+                let (meta, key) = loader.create()?;
+                (meta, key, false)
+            }
+        };
+        let adapter = EncryptingResourceAdapter::<()>::new(key.derive_sub_tree("enc".as_bytes()));
```

Comment:
It seems that both spellings are correct, so we should just pick one to standardize. I think I'll make an issue to address this separately since it's not really related to this PR

Reply:
https://github.com/EspressoSystems/seahorse/issues/245